### ISSUE TITLE
Fix authentication by IP address for NIH

### ISSUE THREAD

**Issue by cmndr-king** [NIHL/Drupal7-Migration#43]
Ensure that general public cannot see the journal access username/password field.

**Comment by tgifford-webfirst** [comment]
Please provide more context--what specifically needs to be fixed? Which content type and field should be restricted?

**Comment by mashanadavis** [comment]
I'll give it a try.  The display of website content---general, training, and electronic resources---is controlled by the taxonomy called Customer Group (http://nihlibrarystg.prod.acquia-sites.com/admin/structure/taxonomy/customer_group). IP addresses are attached to each taxonomy term and each content type has a field called Customer Group (or similar). The user should only see what is assigned to their group, and John commented that there is code that verifies this. He is not sure it is working. 

James' comment about restricting access to the public is related to electronic resources which have not been imported as of yet. And access should go as follows:

NIH - full reign
Public - full reign, except electronic resources (books, journals, databases)
CMS - limited access - CMS only resources and content
HHS - limited access - HHS only resources and content

**Comment by jronmorgan** [comment]
Ted and I just spoke about this. As we discussed these displays are the result of 2 separate but related pieces of logic. The first piece is determining the agency from the incoming IP address. and setting the appropriate Customer Group taxonomy values into a session array. This first is a completely separate task from actually controlling what gets displayed to the user based on the taxonomies that have been set. If either side is broken the user will not see the expected display. 

I will be running a quick test to see if the IP to Agency mapping is working correctly. The code is there, it  *looks* ok, we just need to test it. If that's working then this ticket is purely on the display side. 

If it is broken then we can still test the display side by hard coding the desired agencies into the array and validating that the pages show properly. Being able to test the site as a given agency is critical to the quick development of the site and so we will build a configuration page to allow the current IP  and very possibly the Customer Group to be set by the developer for testing.

**Comment by cmndr-king** [comment]
One further clarification – the general public can see the journals, books, and databases we have available and can link to them but the “Instructions” or “Access Info” field (depending upon which source you look at) should only be shown to NIH IP ranges.  This is more nuanced than turning journals/books/dbs on or off depending upon IP but is down to an individual field.

If this is too hard to do with version one, I’m fine with figuring out a different approach.  We could create a secure page that people link to that includes the username and password since we have so few journals that still do this.

-James
_______________________________
James King, MLS, FSLA
(301) 496-2187

http://about.me/cmndr_king

**Comment by tgifford-webfirst** [comment]
The restrictions on individual fields can be added in a hook_field_access(). 

After John adds tests for the ip address mapping, the performance of that code could also be improved. This is important because page caching has to be disabled for any page that has access controls of either type (node or field). Views on the affected pages may also need to be optimized or selectively cached.

**Comment by tgifford-webfirst** [comment]
For the per-field restrictions, is this specifically about field_access_info (Long Text on the Resource content type)? Do other fields in the associated field group need to be restricted?

field_access_info is not found on the dev site--was that a recent change?

**Comment by jronmorgan** [comment]
Per field_access_info  being changed I would have no idea. WHen we started working on this I copied stg back to dev so it mute be new however its not something I did.  I checked the database and the table field_data_field_access_info has no records.  Maybe it is something new someone was experimenting with?

**Comment by mashanadavis** [comment]
http://nihlibrary.nih.gov/ResearchTools/pages/default.aspx?srchType=OnlineJournals

**Comment by tgifford-webfirst** [comment]
Ah, now it clicks for me. This field-level access control doesn't need to take the user's Agency into account, only whether or not they are "general public", which can be determined solely by IP address.

**Comment by jronmorgan** [comment]
Yes, I spoke with MaShana yesterday to get some more clarification on what this problem was. It seems to have really nothing to do with #44 and is simply a check of IP address and possibly user login status (ie. if the user is logged in any way then he's obviously from some valid source as well). 

suggestion: Rather than checking actual IP every time I might suggest putting it in a session variable so that we can more easily spoof the values for testing. That, or provide a single function for accessing the current IP as opposed to getting it directly from the environment in multiple places in the code.

**Comment by cmndr-king** [comment]
To help be specific, when someone authenticates to the site, either by being from the correct IP range or runs through a Proxy server, they will then be able to see the ERM username/password field which I believe is called "Access Info" in Drupal.  If the user is coming from the general web, they should not see this field.  John requested a link to see the field in action but since we don't have data in the ERM yet, I can't show it.  However, if you create a resource, and click on the "Access Info" tab, the last block is an "Access Info" field.  I believe this field is what is supposed to have access instructions that are to only be visible to authenticated users.

**Comment by mashanadavis** [comment]
I clarified this with John and added screenshots to Waffle. Now the question is how to check/validate whether a user should see the field.

MaShana Davis, MIM
Information Architect
National Institutes of Health Library
Division of Library Services
Office of Research Services
10 Center Drive
Building 10, Room 1L13E, MSC 1150
Bethesda, MD 20892-1150
301.594.6470
mashana.davis@nih.gov<mailto:mashana.davis@nih.gov>

Stay connected with the NIH Library
NIH Library: http://nihlibrary.nih.gov
Facebook: http://www.facebook.com/nihlibrary
Twitter: http://www.twitter.com/nihlib
------------------------------------------
Your Partner in Research




On Oct 18, 2015, at 10:51 PM, James King <notifications@github.com<mailto:notifications@github.com>> wrote:


To help be specific, when someone authenticates to the site, either by being from the correct IP range or runs through a Proxy server, they will then be able to see the ERM username/password field which I believe is called "Access Info" in Drupal. If the user is coming from the general web, they should not see this field. John requested a link to see the field in action but since we don't have data in the ERM yet, I can't show it. However, if you create a resource, and click on the "Access Info" tab, the last block is an "Access Info" field. I believe this field is what is supposed to have access instructions that are to only be visible to authenticated users.

**Comment by cmndr-king** [comment]
That is tied to the other tickets that setup the authentication.  I think we first need to fix those tickets before we can complete this one.


-James
_______________________________
James King, MLS, FSLA
(301) 496-2187

http://about.me/cmndr_king

**Comment by jronmorgan** [comment]
The two are separable as long as we have an interface in place (as we seem to now) for interacting between them. Basically if the set session variables are used for display control then those values will be known and controllable and development can proceed. I believe that we should add the IP address to the SESSION also so that it can be more easily manipulated for testing purposes as opposed to getting it from the environment every time.

**Comment by tgifford-webfirst** [comment]
I added the IP address to the session last week, and fixed an issue with when the authentication code was running. I can move forward on this ticket this afternoon.

**Comment by tgifford-webfirst** [comment]
There is also an "Access Information" field on the tool content type (field_tool_access_info). Does that need to be restricted?

**Comment by mashanadavis** [comment]
No. This field has to do with software located on our workstations. This is not related to the ERM (or “resources” content type). James and/or Ben can confirm.

**Comment by tgifford-webfirst** [comment]
This is now implemented. I created an example Resource that is accessible by the public, except for the access info field: http://nihlibrarystg.prod.acquia-sites.com/content/access-info-test

For an example of a resource that is *not* publicly accessible, but visible based on selected customer groups, see http://nihlibrarystg.prod.acquia-sites.com/node/1026. You will need to be able to set your customer group for testing, details on how to do that will be posted to basecamp. 

Note that the site displays some nodes with the "Admin" display style, which can affect which fields are shown as well. This will not override the field_access restrictions implemented here.

**Comment by jronmorgan** [comment]
What are the actual physical IP addresses in use when the user is on campus at NIH, HHS and CMS?

**Comment by mashanadavis** [comment]
Hi John – Sent in a separate email:
The IP addresses are included in each customer group profile: http://nihlibrarystg.prod.acquia-sites.com/admin/structure/taxonomy/customer_group.

Not sure if this was updated on DEV.

**Comment by cmndr-king** [comment]
Marking this as closed.
